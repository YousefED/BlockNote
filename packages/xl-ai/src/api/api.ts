import {
  Block,
  BlockNoteEditor,
  BlockSchema,
  InlineContentSchema,
  StyleSchema,
} from "@blocknote/core";
import {
  CoreMessage,
  LanguageModel,
  StreamObjectResult,
  generateText,
  jsonSchema,
  streamObject,
} from "ai";
import { markdownNodeDiff } from "../markdown/markdownNodeDiff.js";
import { markdownNodeDiffToBlockOperations } from "../markdown/markdownOperations.js";
import { addFunction } from "./functions/add.js";
import { deleteFunction } from "./functions/delete.js";
import { AIFunction } from "./functions/index.js";
import { updateFunction } from "./functions/update.js";
import { createOperationsArraySchema } from "./schema/operations.js";
import { blockNoteSchemaToJSONSchema } from "./schema/schemaToJSONSchema.js";
import { trimArray } from "./util/trimArray.js";

// TODO don't include child block
export function createMessagesForLLM(opts: {
  editor: BlockNoteEditor;
  prompt: string;
  document: any;
}): Array<CoreMessage> {
  if (opts.editor.getSelection()) {
    return [
      {
        role: "system",
        content: `You're manipulating a text document. Make sure to follow the json schema provided. 
          The user selected everything between [$! and !$], including blocks in between.`,
      },
      {
        role: "system",
        content: JSON.stringify(
          suffixIDs(opts.editor.getSelectionWithMarkers() as any)
        ),
      },
      {
        role: "system",
        content:
          "Make sure to ONLY affect the selected text and blocks (split words if necessary), and don't include the markers in the response.",
      },
      {
        role: "user",
        content: opts.prompt,
      },
    ];
  }
  return [
    {
      role: "system",
      content:
        "You're manipulating a text document. Make sure to follow the json schema provided. This is the document:",
    },
    {
      role: "system",
      content: JSON.stringify(suffixIDs(opts.document)),
    },
    {
      role: "system",
      content:
        "This would be an example block: \n" +
        JSON.stringify({
          type: "paragraph",
          props: {},
          content: [
            {
              type: "text",
              text: "Bold text",
              styles: {
                bold: true,
              },
            },
            {
              type: "text",
              text: " and italic text",
              styles: {
                italic: true,
              },
            },
          ],
        }),
    },
    {
      role: "user",
      content: opts.prompt,
    },
  ];
}

export function suffixIDs<
  T extends Block<BlockSchema, InlineContentSchema, StyleSchema>
>(blocks: T[]): T[] {
  return blocks.map((block) => ({
    ...block,
    id: `${block.id}$`,
    children: suffixIDs(block.children),
  }));
}

type CallLLMOptions = {
  model: LanguageModel;
  functions?: AIFunction[];
} & (
  | {
      prompt: string;
    }
  | {
      messages: Array<CoreMessage>;
    }
);

export async function callLLMStreamingMD(
  editor: BlockNoteEditor<any, any, any>,
  options: CallLLMOptions & {
    _streamObjectOptions?: Partial<Parameters<typeof streamObject<any>>[0]>;
  }
) {
  const markdown = await editor.blocksToMarkdownLossy();

  const withDefaults: Required<
    Omit<CallLLMOptions, "prompt"> & { messages: CoreMessage[] }
  > = {
    messages: [
      {
        role: "system",
        content:
          "You're manipulating a markdown document. Send me the updated markdown. Existing document:",
      },
      {
        role: "system",
        content: markdown,
      },
      {
        role: "user",
        content: (options as any).prompt,
      },
    ],
    ...options,
  };

  const ret = await generateText<any>({
    model: withDefaults.model,
    messages: withDefaults.messages,
    ...(options._streamObjectOptions as any),
  });

  const blocks = trimArray(editor.document, (block) => {
    return (
      block.type === "paragraph" &&
      Array.isArray(block.content) &&
      block.content.length === 0
    );
  });
  const newMarkdown = ret.text.trim();

  const diff = await markdownNodeDiff(markdown, newMarkdown);
  const operations = await markdownNodeDiffToBlockOperations(
    editor,
    blocks,
    diff
  );

  for (const operation of operations) {
    await applyAIOperation(
      operation,
      editor,
      [updateFunction, addFunction, deleteFunction],
      undefined,
      { idsSuffixed: false }
    );
  }
}

export async function callLLMStreaming(
  editor: BlockNoteEditor<any, any, any>,
  options: CallLLMOptions & {
    _streamObjectOptions?: Partial<Parameters<typeof streamObject<any>>[0]>;
  }
) {
  const withDefaults: Required<
    Omit<CallLLMOptions, "prompt"> & { messages: CoreMessage[] }
  > = {
    functions: [updateFunction, addFunction, deleteFunction],
    messages:
      (options as any).messages ||
      createMessagesForLLM({
        editor,
        prompt: (options as any).prompt,
        document: editor.document,
      }),

    ...options,
  };

  const ret = streamObject<any>({
    model: withDefaults.model,
    mode: "tool",
    schema: jsonSchema({
      ...createOperationsArraySchema(withDefaults.functions),
      $defs: blockNoteSchemaToJSONSchema(editor.schema).$defs as any,
    }),
    messages: withDefaults.messages,
    ...(options._streamObjectOptions as any),
  });
  return await applyLLMResponse(editor, ret, withDefaults.functions);
}

export function applyAIOperation(
  operation: any,
  editor: BlockNoteEditor,
  functions: AIFunction[],
  operationContext: any,
  options: {
    idsSuffixed: boolean;
  } = {
    idsSuffixed: false,
  }
) {
  const func = functions.find((func) => func.schema.name === operation.type);
  if (!func || !func.validate(operation, editor, options)) {
    console.log("INVALID OPERATION", operation);
    return operationContext;
  }
  return func.apply(operation, editor, operationContext, options);
}

export async function applyLLMResponse(
  editor: BlockNoteEditor,
  response: StreamObjectResult<any, any, any>,
  functions: AIFunction[]
) {
  let numOperationsAppliedCompletely = 0;
  let operationContext: any = undefined;

  for await (const partialObject of response.partialObjectStream) {
    const operations: [] = partialObject.operations || [];
    console.log(operations);
    let isFirst = true;
    for (const operation of operations.slice(numOperationsAppliedCompletely)) {
      operationContext = applyAIOperation(
        operation,
        editor,
        functions,
        isFirst ? operationContext : undefined
      );
      isFirst = false;
    }

    numOperationsAppliedCompletely = operations.length - 1;
  }
  return response;
}

// - cursor position
// - API design (customize context, cursor position, prompt, stream / nostream, validation)
