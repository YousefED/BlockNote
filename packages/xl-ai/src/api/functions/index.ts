import { addFunction } from "./add.js";
import { deleteFunction } from "./delete.js";
import { updateFunction } from "./update.js";

export type AIFunction =
  | typeof addFunction
  | typeof deleteFunction
  | typeof updateFunction;
