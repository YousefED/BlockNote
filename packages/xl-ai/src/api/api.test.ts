import { afterEach, describe, expect, it } from "vitest";

import { BlockNoteEditor, PartialBlock } from "@blocknote/core";

import { createGroq } from "@ai-sdk/groq";
import { createOpenAI } from "@ai-sdk/openai";
import { callLLMStreaming } from "./api.js";
import { createBlockNoteAIClient } from "./client.js";

function createEditor(initialContent: PartialBlock[]) {
  return BlockNoteEditor.create({
    initialContent,
  });
}

const client = createBlockNoteAIClient({
  baseURL: "https://localhost:3000/ai",
  apiKey: "PLACEHOLDER",
});

const groq = createGroq({
  ...client.getProviderSettings("groq"),
})("llama-3.1-70b-versatile");

const openai = createOpenAI({
  ...client.getProviderSettings("openai"),
})("gpt-4o-2024-08-06", {});

describe("Test AI operations", () => {
  afterEach(() => {
    delete (window as Window & { __TEST_OPTIONS?: any }).__TEST_OPTIONS;
  });

  describe("Update", () => {
    it("translates simple paragraphs", async () => {
      const editor = createEditor([
        {
          type: "paragraph",
          content: "Hello",
        },
        {
          type: "paragraph",
          content: "World",
        },
      ]);

      const response = await callLLMStreaming(editor, {
        model: groq,
        prompt: "translate to german",
      });

      // Add assertions here to check if the document was correctly translated
      // For example:
      expect(editor.document).toMatchSnapshot();

      // expect(await response.object).toMatchSnapshot();
    });

    it("changes simple formatting (paragraph)", async () => {
      const editor = createEditor([
        {
          type: "paragraph",
          content: "Hello",
        },
        {
          type: "paragraph",
          content: "World",
        },
      ]);

      const response = await callLLMStreaming(editor, {
        prompt: "change first paragraph to bold",
      });

      // Add assertions here to check if the document was correctly translated
      // For example:
      expect(editor.document).toMatchSnapshot();

      // expect(await response.object).toMatchSnapshot();
    });

    it("changes simple formatting (word)", async () => {
      const editor = createEditor([
        {
          type: "paragraph",
          content: "Hello world",
        },
      ]);

      const response = await callLLMStreaming(editor, {
        prompt: "change first word to bold",
      });

      // Add assertions here to check if the document was correctly translated
      // For example:
      expect(editor.document).toMatchSnapshot();

      // expect(await response.object).toMatchSnapshot();
    });
  });

  describe("Delete", () => {
    it("deletes a paragraph", async () => {
      const editor = createEditor([
        {
          type: "paragraph",
          content: "Hello",
        },
        {
          type: "paragraph",
          content: "World",
        },
      ]);
      const response = await callLLMStreaming(editor, {
        prompt: "delete the first sentence",
      });

      expect(editor.document).toMatchSnapshot();

      // expect(await response.object).toMatchSnapshot();
    });
  });

  describe("Insert", () => {
    it("inserts a paragraph at start", async () => {
      const editor = createEditor([
        {
          type: "paragraph",
          content: "Hello",
        },
      ]);
      const response = await callLLMStreaming(editor, {
        prompt: "Add a sentence with `Test` before the first sentence",
      });

      expect(editor.document).toMatchSnapshot();

      // expect(await response.object).toMatchSnapshot();
    });

    it("inserts a paragraph at end", async () => {
      const editor = createEditor([
        {
          type: "paragraph",
          content: "Hello",
        },
      ]);
      const response = await callLLMStreaming(editor, {
        prompt: "Add a sentence with `Test` after the first sentence",
      });

      expect(editor.document).toMatchSnapshot();

      // expect(await response.object).toMatchSnapshot();
    });
  });
});
