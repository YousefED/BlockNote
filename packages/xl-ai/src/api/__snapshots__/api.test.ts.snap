// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test AI operations > Delete > deletes a paragraph 1`] = `
[
  {
    "children": [],
    "content": [
      {
        "styles": {},
        "text": "World",
        "type": "text",
      },
    ],
    "id": "1",
    "props": {
      "backgroundColor": "default",
      "textAlignment": "left",
      "textColor": "default",
    },
    "type": "paragraph",
  },
]
`;

exports[`Test AI operations > Delete > deletes a paragraph 2`] = `
{
  "operations": [
    {
      "id": "0$",
      "type": "delete",
    },
  ],
}
`;

exports[`Test AI operations > Insert > inserts a paragraph at end 1`] = `
[
  {
    "children": [],
    "content": [
      {
        "styles": {},
        "text": "Hello",
        "type": "text",
      },
    ],
    "id": "0",
    "props": {
      "backgroundColor": "default",
      "textAlignment": "left",
      "textColor": "default",
    },
    "type": "paragraph",
  },
  {
    "children": [],
    "content": [
      {
        "styles": {},
        "text": "Test",
        "type": "text",
      },
    ],
    "id": "1",
    "props": {
      "backgroundColor": "default",
      "textAlignment": "left",
      "textColor": "default",
    },
    "type": "paragraph",
  },
]
`;

exports[`Test AI operations > Insert > inserts a paragraph at end 2`] = `
{
  "operations": [
    {
      "blocks": [
        {
          "content": [
            {
              "styles": {},
              "text": "Test",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
      "position": "after",
      "referenceId": "0$",
      "type": "add",
    },
  ],
}
`;

exports[`Test AI operations > Insert > inserts a paragraph at start 1`] = `
[
  {
    "children": [],
    "content": [
      {
        "styles": {},
        "text": "Test",
        "type": "text",
      },
    ],
    "id": "1",
    "props": {
      "backgroundColor": "default",
      "textAlignment": "left",
      "textColor": "default",
    },
    "type": "paragraph",
  },
  {
    "children": [],
    "content": [
      {
        "styles": {},
        "text": "Hello",
        "type": "text",
      },
    ],
    "id": "0",
    "props": {
      "backgroundColor": "default",
      "textAlignment": "left",
      "textColor": "default",
    },
    "type": "paragraph",
  },
]
`;

exports[`Test AI operations > Insert > inserts a paragraph at start 2`] = `
{
  "operations": [
    {
      "blocks": [
        {
          "content": [
            {
              "styles": {},
              "text": "Test",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
      "position": "before",
      "referenceId": "0$",
      "type": "add",
    },
  ],
}
`;

exports[`Test AI operations > Update > changes simple formatting (paragraph) 1`] = `
[
  {
    "children": [],
    "content": [
      {
        "styles": {
          "bold": true,
        },
        "text": "Hello",
        "type": "text",
      },
    ],
    "id": "0",
    "props": {
      "backgroundColor": "default",
      "textAlignment": "left",
      "textColor": "default",
    },
    "type": "paragraph",
  },
  {
    "children": [],
    "content": [
      {
        "styles": {},
        "text": "World",
        "type": "text",
      },
    ],
    "id": "1",
    "props": {
      "backgroundColor": "default",
      "textAlignment": "left",
      "textColor": "default",
    },
    "type": "paragraph",
  },
]
`;

exports[`Test AI operations > Update > changes simple formatting (paragraph) 2`] = `
{
  "operations": [
    {
      "block": {
        "content": [
          {
            "styles": {
              "bold": true,
            },
            "text": "Hello",
            "type": "text",
          },
        ],
        "type": "paragraph",
      },
      "id": "0$",
      "type": "update",
    },
  ],
}
`;

exports[`Test AI operations > Update > changes simple formatting (word) 1`] = `
[
  {
    "children": [],
    "content": [
      {
        "styles": {
          "bold": true,
        },
        "text": "Hello",
        "type": "text",
      },
      {
        "styles": {},
        "text": " world",
        "type": "text",
      },
    ],
    "id": "0",
    "props": {
      "backgroundColor": "default",
      "textAlignment": "left",
      "textColor": "default",
    },
    "type": "paragraph",
  },
]
`;

exports[`Test AI operations > Update > changes simple formatting (word) 2`] = `
{
  "operations": [
    {
      "block": {
        "content": [
          {
            "styles": {
              "bold": true,
            },
            "text": "Hello",
            "type": "text",
          },
          {
            "styles": {},
            "text": " world",
            "type": "text",
          },
        ],
        "type": "paragraph",
      },
      "id": "0$",
      "type": "update",
    },
  ],
}
`;

exports[`Test AI operations > Update > translates simple paragraphs 1`] = `
[
  {
    "children": [],
    "content": [
      {
        "styles": {},
        "text": "Hallo",
        "type": "text",
      },
    ],
    "id": "0",
    "props": {
      "backgroundColor": "default",
      "textAlignment": "left",
      "textColor": "default",
    },
    "type": "paragraph",
  },
  {
    "children": [],
    "content": [
      {
        "styles": {},
        "text": "Welt",
        "type": "text",
      },
    ],
    "id": "1",
    "props": {
      "backgroundColor": "default",
      "textAlignment": "left",
      "textColor": "default",
    },
    "type": "paragraph",
  },
]
`;

exports[`Test AI operations > Update > translates simple paragraphs 2`] = `
{
  "operations": [
    {
      "block": {
        "content": [
          {
            "styles": {},
            "text": "Hallo",
            "type": "text",
          },
        ],
        "type": "paragraph",
      },
      "id": "0$",
      "type": "update",
    },
    {
      "block": {
        "content": [
          {
            "styles": {},
            "text": "Welt",
            "type": "text",
          },
        ],
        "type": "paragraph",
      },
      "id": "1$",
      "type": "update",
    },
  ],
}
`;

exports[`Test AI operations > changes simple formatting (paragraph) 1`] = `
[
  {
    "children": [],
    "content": [
      {
        "styles": {
          "bold": true,
        },
        "text": "Hello",
        "type": "text",
      },
    ],
    "id": "0",
    "props": {
      "backgroundColor": "default",
      "textAlignment": "left",
      "textColor": "default",
    },
    "type": "paragraph",
  },
  {
    "children": [],
    "content": [
      {
        "styles": {},
        "text": "World",
        "type": "text",
      },
    ],
    "id": "1",
    "props": {
      "backgroundColor": "default",
      "textAlignment": "left",
      "textColor": "default",
    },
    "type": "paragraph",
  },
]
`;

exports[`Test AI operations > changes simple formatting (word) 1`] = `
[
  {
    "children": [],
    "content": [
      {
        "styles": {
          "bold": true,
        },
        "text": "Hello",
        "type": "text",
      },
      {
        "styles": {},
        "text": " world",
        "type": "text",
      },
    ],
    "id": "0",
    "props": {
      "backgroundColor": "default",
      "textAlignment": "left",
      "textColor": "default",
    },
    "type": "paragraph",
  },
]
`;

exports[`Test AI operations > translates simple paragraphs 1`] = `
[
  {
    "children": [],
    "content": [
      {
        "styles": {},
        "text": "Hallo",
        "type": "text",
      },
    ],
    "id": "0",
    "props": {
      "backgroundColor": "default",
      "textAlignment": "left",
      "textColor": "default",
    },
    "type": "paragraph",
  },
  {
    "children": [],
    "content": [
      {
        "styles": {},
        "text": "Welt",
        "type": "text",
      },
    ],
    "id": "1",
    "props": {
      "backgroundColor": "default",
      "textAlignment": "left",
      "textColor": "default",
    },
    "type": "paragraph",
  },
]
`;
